pallets/phala/src/stakepool.rs

```diff
@@ -1110,11 +1032,181 @@ pub mod pallet {
+			let amount = bmul(shares, &price);
+
+			let collection_id = PoolCollections::<T>::get(pool_info.pid.clone())?;
+			let nft_id =
+				Self::merge_or_init_nft_for_staker(userid.clone(), pool_info.pid.clone()).ok()?;
+			let mut nft = Self::get_NftAttr(collection_id, nft_id).ok()?;
+
+			let amount = amount.min(pool_info.free_stake).min(nft.stakes);
+
+			let user_shares = nft.shares.checked_sub(&shares)?;
+			let (user_shares, shares_dust) = extract_dust(user_shares);
+			let user_locked = nft.stakes.checked_sub(&amount)?;
+			let (user_locked, user_dust) = extract_dust(user_locked);
+
+			let removed_shares = shares + shares_dust;
+			let total_shares = pool_info.total_shares.checked_sub(&removed_shares)?;
+
+			let (total_stake, _) = extract_dust(pool_info.total_stake - amount);
+			if total_stake > Zero::zero() {
+				pool_info.free_stake -= amount;
+				pool_info.total_stake -= amount;
+			} else {
+				pool_info.free_stake = Zero::zero();
+				pool_info.total_stake = Zero::zero();
+			}
+			pool_info.total_shares = total_shares;
+			nft.shares = user_shares;
+			nft.stakes = user_locked;
+
+			Some((amount, user_dust, removed_shares))
+		}
+
+		#[frame_support::transactional]
+		fn mint_and_lock_pool_nft(
+			pid: u64,
+			contributer: T::AccountId,
+			shares: BalanceOf<T>,
+			stakes: BalanceOf<T>,
+		) -> Result<NftId, DispatchError> {
+			let pool_info = Self::ensure_pool(pid)?;
+			let collection_id =
+				PoolCollections::<T>::get(pid).ok_or(Error::<T>::MissCollectionId)?;
+			let collection_info = pallet_rmrk_core::Collections::<T>::get(collection_id)
+				.ok_or(pallet_rmrk_core::Error::<T>::CollectionUnknown)?;
+			let nft_id = pallet_rmrk_core::NextNftId::<T>::get(collection_id);
+
+			pallet_rmrk_core::Pallet::<T>::mint_nft(
+				Origin::<T>::Signed(pool_info.owner.clone()).into(),
+				contributer,
+				collection_id,
+				None,
+				None,
+				collection_info.metadata.clone(),
+				true,
+			)?;
+
+			let attr = NftAttr { shares, stakes };
+			Self::set_NftAttr(pid, collection_id, nft_id, &attr)?;
+			pallet_rmrk_core::Pallet::<T>::set_lock((collection_id, nft_id), true);
+			Ok(nft_id)
+		}
+
+		#[frame_support::transactional]
+		fn unlock_and_burn_nft(pid: u64, nft_id: NftId) -> DispatchResult {
+			let pool_info = Self::ensure_pool(pid)?;
+			let collection_id =
+				PoolCollections::<T>::get(pid).ok_or(Error::<T>::MissCollectionId)?;
+			pallet_rmrk_core::Pallet::<T>::set_lock((collection_id, nft_id), false);
+			pallet_rmrk_core::Pallet::<T>::burn_nft(
+				Origin::<T>::Signed(pool_info.owner.clone()).into(),
+				collection_id,
+				nft_id,
+			)?;
+
+			Ok(())
+		}
+
+		#[frame_support::transactional]
+		fn merge_or_init_nft_for_staker(
+			staker: T::AccountId,
+			pid: u64,
+		) -> Result<NftId, DispatchError> {
+			let pool_info = Self::ensure_pool(pid)?;
+			let collection_id =
+				PoolCollections::<T>::get(pid).ok_or(Error::<T>::MissCollectionId)?;
+			let nftid_arr: Vec<NftId> =
```

Comment:
Iterate with this: https://github.com/paritytech/substrate/blob/0fbf917676824cda9c3cf7e71e83b20c26b5f92e/frame/uniques/src/lib.rs#L173-L182

Reply:
waiting for joshua's reply